File: frontend/src/api/mockData.js
const agents = ['planner', 'search', 'analysis', 'writer', 'reviewer']

function isoMinutes(offset = 0) {
  const date = new Date(Date.now() + offset * 60 * 1000)
  date.setSeconds(0, 0)
  return date.toISOString().slice(0, 16)
}

export function mockOverview() {
  return {
    running_tasks: 18,
    active_agents: 5,
    events_per_minute: 126,
    success_rate: 0.934,
    error_rate: 0.066,
    avg_latency_ms: 1840,
    token_total_5m: 48230,
    estimated_cost_5m: 0.0832,
    retry_tasks: 3,
    open_alerts: 4,
    updated_at: new Date().toISOString().slice(0, 19)
  }
}

export function mockTrend(minutes = 60) {
  return Array.from({ length: Math.min(minutes, 60) }, (_, index) => ({
    time: isoMinutes(index - Math.min(minutes, 60) + 1),
    events: 80 + (index * 7) % 55,
    success: 70 + (index * 5) % 45,
    failed: 3 + index % 8,
    avg_latency_ms: 1200 + (index * 113) % 2200
  }))
}

export function mockAgents() {
  return agents.map((role, index) => ({
    agent_id: `${role}_agent`,
    agent_role: role,
    status: ['running', 'running', 'idle', 'running', 'retry'][index],
    current_task: `trace_demo_${String(index + 1).padStart(3, '0')}`,
    success_rate: Number((0.94 - index * 0.013).toFixed(3)),
    avg_latency_ms: 900 + index * 430,
    token_total: 12000 + index * 6800,
    retry_count: index % 3,
    last_event_time: new Date().toISOString().slice(0, 19)
  }))
}

export function mockRealtimeAlerts() {
  const now = new Date()
  return [
    {
      alert_id: 'alert_demo_latency',
      alert_type: 'high_latency',
      level: 'warning',
      agent_id: 'writer_agent',
      current_value: 12800,
      threshold: 10000,
      source: 'streaming',
      status: 'open',
      created_at: new Date(now.getTime() - 3 * 60 * 1000).toISOString().slice(0, 19)
    },
    {
      alert_id: 'alert_demo_retry',
      alert_type: 'frequent_retry',
      level: 'critical',
      agent_id: 'reviewer_agent',
      current_value: 4,
      threshold: 3,
      source: 'streaming',
      status: 'open',
      created_at: new Date(now.getTime() - 6 * 60 * 1000).toISOString().slice(0, 19)
    },
    {
      alert_id: 'alert_demo_recovery',
      alert_type: 'stream_recovery',
      level: 'info',
      agent_id: 'spark_streaming',
      current_value: 1,
      threshold: 1,
      source: 'streaming',
      status: 'open',
      created_at: new Date(now.getTime() - 9 * 60 * 1000).toISOString().slice(0, 19)
    }
  ]
}

export function mockDailyMetrics(startDate = '2026-06-01', endDate = '2026-06-24') {
  const start = new Date(`${startDate}T00:00:00`)
  const end = new Date(`${endDate}T00:00:00`)
  const days = Math.max(1, Math.round((end - start) / 86400000) + 1)
  return Array.from({ length: days }, (_, index) => {
    const date = new Date(start.getTime() + index * 86400000).toISOString().slice(0, 10)
    return {
      metric_date: date,
      task_count: 1000 + index * 82,
      success_count: 920 + index * 75,
      failed_count: 80 + index * 7,
      success_rate: Number((0.91 + (index % 5) * 0.008).toFixed(3)),
      avg_latency_ms: 1600 + index * 60,
      p95_latency_ms: 4100 + index * 90,
      total_tokens: 1200000 + index * 95000,
      estimated_cost_usd: Number((2.4 + index * 0.18).toFixed(4))
    }
  })
}

export function mockHourlyMetrics() {
  return Array.from({ length: 24 }, (_, hour) => ({
    hour,
    task_count: 36 + hour * 2,
    success_rate: Number((0.9 + (hour % 6) * 0.01).toFixed(3)),
    avg_latency_ms: 1100 + (hour * 137) % 1800,
    total_tokens: 30000 + hour * 4200
  }))
}

export function mockRankings(date = '2026-06-24') {
  return agents.map((role, index) => ({
    metric_date: date,
    agent_id: `${role}_agent`,
    agent_role: role,
    execution_count: 1200 - index * 95,
    success_rate: Number((0.94 - index * 0.015).toFixed(3)),
    avg_latency_ms: 980 + index * 520,
    p95_latency_ms: 2600 + index * 900,
    total_tokens: 210000 + index * 78000,
    estimated_cost_usd: Number((0.48 + index * 0.17).toFixed(4))
  }))
}

export function mockRelationGraph(date = '2026-06-24') {
  return {
    metric_date: date,
    nodes: agents.map((role, index) => ({ id: `${role}_agent`, name: role[0].toUpperCase() + role.slice(1), value: 100 - index * 9 })),
    links: agents.slice(0, -1).map((role, index) => ({
      source: `${role}_agent`,
      target: `${agents[index + 1]}_agent`,
      call_count: 900 - index * 70,
      avg_latency_ms: 1200 + index * 330,
      failed_count: index * 8,
      total_tokens: 80000 + index * 32000
    }))
  }
}

export function mockHistoryAlerts(date = '2026-06-24') {
  return [
    ['high_error_rate', 'warning', 'search_agent', 0.24, 0.2],
    ['high_p95_latency', 'warning', 'writer_agent', 12400, 10000],
    ['high_cost_task', 'info', 'analysis_agent', 0.72, 0.5]
  ].map(([alertType, level, agentId, currentValue, threshold], index) => ({
    alert_id: `hist_${date}_${index}`,
    metric_date: date,
    alert_type: alertType,
    level,
    agent_id: agentId,
    current_value: currentValue,
    threshold,
    source: 'batch',
    status: 'open'
  }))
}

export function mockReport(payload = {}) {
  const date = payload.report_date || '2026-06-24'
  return {
    report_id: `report_${date}`,
    report_date: date,
    report_type: payload.report_type || 'daily',
    created_at: new Date().toISOString().slice(0, 19),
    content: `# AgentScope ${date} 运行分析报告

## 总体结论

系统实时链路和离线分析链路均处于可观测状态。Planner、Search、Analysis、Writer、Reviewer 五类 Agent 均产生有效事件，整体成功率保持在 93% 左右。

## 主要异常

- Writer Agent 出现高延迟告警，P95 时延超过阈值。
- Reviewer Agent 存在频繁重试现象，需要关注任务回退逻辑。
- Search Agent 错误率略高，建议检查外部检索工具稳定性。

## 优化建议

1. 对 Writer Agent 增加超时保护和降级策略。
2. 对 Reviewer 重试次数设置硬阈值，避免疑似循环调用。
3. 将实时告警和历史告警统一纳入 AI 报告摘要。`
  }
}

export function mockReports() {
  return [mockReport({ report_date: '2026-06-24', report_type: 'daily' })]
}

File: backend/app/services/mock_data.py
"""
mock_data.py —— AgentScope 数据指标模拟生成层 (Fallback / Mock Data Provider)

在系统尚未启动离线仿真、或者是后端组件（如 Kafka, MySQL）断开连接时，
本模块作为兜底方案，利用系统当前秒数和配置基准，动态产生以下场景的 Mock 仿真数据：
  1. 实时数据概览 (realtime_overview)
  2. 实时历史趋势 (realtime_trend)
  3. 实时智能体运行状态 (realtime_agents)
  4. 实时告警积压记录 (recent_alerts)
  5. 离线每日核心多维指标 (daily_metrics)
  6. 离线单日小时细粒度指标 (hourly_metrics)
  7. 单日 Agent 指标效能排行榜 (rankings)
  8. 调用关系网络拓扑图 (relation_graph)
  9. 批量历史告警数据 (history_alerts)
"""

from __future__ import annotations

from datetime import date, datetime, timedelta
from typing import Dict, List


# 智能体标准角色清单
AGENTS = ["planner", "search", "analysis", "writer", "reviewer"]


def realtime_overview() -> Dict:
    """
    动态生成实时监控中心顶部的核心概览卡片数据。
    使用当前物理秒数 (datetime.now().second) 做取模运算，使数据在每次刷新时具备微弱波动性，营造真实演进的视觉效果。

    Returns:
        Dict: 包含运行中任务数、吞吐量、成功率、时延、消耗成本等度量指标的字典
    """
    sec = datetime.now().second
    return {
        "running_tasks": 15 + sec % 10,                 # 模拟 15-24 个并发执行的任务
        "active_agents": 5,                             # 固定 5 类智能体
        "events_per_minute": 120 + sec % 20,            # 模拟每分钟事件处理速率
        "success_rate": 0.92 + (sec % 5) * 0.01,        # 成功率在 92% ~ 96% 间抖动
        "error_rate": 0.08 - (sec % 5) * 0.01,          # 错误率与成功率互补
        "avg_latency_ms": 1500 + (sec % 15) * 100,      # 时延在 1.5s ~ 2.9s 之间波动
        "token_total_5m": 45000 + sec * 200,            # 最近 5 分钟 Token 消费累计量
        "estimated_cost_5m": 0.08 + (sec % 10) * 0.002, # 最近 5 分钟估算花费
        "retry_tasks": sec % 4,                         # 重试队列积压数
        "open_alerts": sec % 3,                         # 当前未解决的告警条数
        "update_time": datetime.now().isoformat(timespec="seconds"), # 本次渲染的时间截断
    }


def realtime_trend(minutes: int) -> List[Dict]:
    """
    生成过去一段时间内（比如最近 30 秒/分钟）的实时数据处理吞吐与耗时趋势点。

    Args:
        minutes (int): 指定需要生成的趋势点样本长度

    Returns:
        List[Dict]: 趋势数据列表，每个元素包含时间戳、事件数、成功数、失败数及平均耗时
    """
    seconds = minutes
    now = datetime.now()
    rows = []
    for idx in range(seconds):
        ts = now - timedelta(seconds=seconds - idx - 1)
        rows.append(
            {
                "time": ts.isoformat(timespec="seconds"),
                "events": 80 + (idx * 7) % 55,
                "success": 70 + (idx * 5) % 45,
                "failed": 3 + idx % 8,
                "avg_latency_ms": 1200 + (idx * 113) % 2200,
            }
        )
    return rows


def realtime_agents() -> List[Dict]:
    """
    生成当前各类型 Agent 的工作状态、实时成功率、累计 Token 及最新活跃时间。

    Returns:
        List[Dict]: 每个角色的实时度量字典列表
    """
    sec = datetime.now().second
    return [
        {
            "agent_id": f"{role}_agent",
            "agent_role": role,
            # 状态通过秒数动态切换，展示出 Agent 繁忙与闲置的更迭
            "status": "running" if (idx + sec // 5) % 2 == 0 else "idle",
            "current_task": f"trace_demo_{100 + (sec // 10):03d}",
            "success_rate": round(0.91 + ((idx + sec) % 7) * 0.012, 3),
            "avg_latency_ms": 900 + ((idx + sec) % 5) * 430,
            "token_total": 12000 + ((idx + sec) % 20) * 1800,
            "retry_count": (idx + sec // 15) % 3,
            "last_event_time": datetime.now().isoformat(timespec="seconds"),
        }
        for idx, role in enumerate(AGENTS)
    ]


def recent_alerts() -> List[Dict]:
    """
    生成实时告警大屏所需显示的异常告警事件（如：延迟超限、触发高频重试、限流等）。
    通过秒数在 15~45 秒内动态添加额外的严重告警项，使大屏幕在没有真实异常时也能产生交替告警。

    Returns:
        List[Dict]: 活动中的告警项列表
    """
    now = datetime.now()
    sec = now.second
    alerts = [
        {
            "alert_id": "alert_demo_latency",
            "alert_type": "high_latency",
            "level": "warning",
            "agent_id": "writer_agent",
            "current_value": 12800,
            "threshold": 10000,
            "source": "streaming",
            "status": "open",
            "create_time": (now - timedelta(minutes=3)).isoformat(timespec="seconds"),
        },
        {
            "alert_id": "alert_demo_retry",
            "alert_type": "frequent_retry",
            "level": "critical",
            "agent_id": "reviewer_agent",
            "current_value": 4,
            "threshold": 3,
            "source": "streaming",
            "status": "open",
            "create_time": (now - timedelta(minutes=6)).isoformat(timespec="seconds"),
        },
    ]

    # 当秒数在 [15, 45] 之间时，模拟注入大模型 HTTP 429 限流告警
    if 15 <= sec <= 45:
        alerts.append({
            "alert_id": "alert_demo_rate_limit",
            "alert_type": "rate_limit_exceeded",
            "level": "critical",
            "agent_id": "planner_agent",
            "current_value": 429,
            "threshold": 200,
            "source": "streaming",
            "status": "open",
            "create_time": (now - timedelta(seconds=sec - 15)).isoformat(timespec="seconds"),
        })

    # 当秒数大于 30 秒时，模拟注入 Token 溢出告警
    if sec > 30:
        alerts.append({
            "alert_id": "alert_demo_token_overflow",
            "alert_type": "token_limit",
            "level": "warning",
            "agent_id": "search_agent",
            "current_value": 8500,
            "threshold": 8000,
            "source": "streaming",
            "status": "open",
            "create_time": (now - timedelta(seconds=sec - 30)).isoformat(timespec="seconds"),
        })

    return alerts


def daily_metrics(start: date, end: date) -> List[Dict]:
    """
    提供每日离线数仓 ADS 级指标的 Mock 数据（含任务总量、成功率、时延、消耗成本等趋势）。

    Args:
        start (date): 查询起始日期
        end (date): 查询截止日期

    Returns:
        List[Dict]: 按天平铺的历史离线指标数据
    """
    days = (end - start).days + 1
    return [
        {
            "metric_date": (start + timedelta(days=idx)).isoformat(),
            "task_count": 1000 + idx * 82,
            "success_count": 920 + idx * 75,
            "failed_count": 80 + idx * 7,
            "success_rate": round(0.91 + (idx % 5) * 0.008, 3),
            "avg_latency_ms": 1600 + idx * 60,
            "p95_latency_ms": 4100 + idx * 90,
            "total_tokens": 1_200_000 + idx * 95_000,
            "estimated_cost_usd": round(2.4 + idx * 0.18, 4),
        }
        for idx in range(max(days, 0))
    ]


def hourly_metrics(metric_date: date) -> List[Dict]:
    """
    生成单日内以小时为维度的流量分布数据。

    Args:
        metric_date (date): 选定的业务日期

    Returns:
        List[Dict]: 一天 24 小时的指标序列
    """
    return [
        {
            "metric_date": metric_date.isoformat(),
            "hour": hour,
            "task_count": 36 + hour * 2,
            "success_rate": round(0.90 + (hour % 6) * 0.01, 3),
            "avg_latency_ms": 1100 + (hour * 137) % 1800,
            "total_tokens": 30000 + hour * 4200,
        }
        for hour in range(24)
    ]


def rankings(metric_date: date) -> List[Dict]:
    """
    生成指定日期的 Agent 执行效能排行榜单（按执行次数、Token、时延及花费）。

    Args:
        metric_date (date): 业务日期

    Returns:
        List[Dict]: 各 Agent 角色效能字典列表
    """
    return [
        {
            "metric_date": metric_date.isoformat(),
            "agent_id": f"{role}_agent",
            "agent_role": role,
            "execution_count": 1200 - idx * 95,
            "success_rate": round(0.94 - idx * 0.015, 3),
            "avg_latency_ms": 980 + idx * 520,
            "p95_latency_ms": 2600 + idx * 900,
            "total_tokens": 210000 + idx * 78000,
            "estimated_cost_usd": round(0.48 + idx * 0.17, 4),
        }
        for idx, role in enumerate(AGENTS)
    ]


def relation_graph(metric_date: date) -> Dict:
    """
    生成指定日期的智能体调用关系网路拓扑数据（包含节点及连线关系流量）。

    Args:
        metric_date (date): 业务日期

    Returns:
        Dict: 含有 nodes（节点列表）与 links（带权重边列表）的拓扑关系字典
    """
    nodes = [
        {"id": f"{role}_agent", "name": role.title(), "value": 100 - idx * 9}
        for idx, role in enumerate(AGENTS)
    ]
    links = [
        {
            "source": f"{AGENTS[idx]}_agent",
            "target": f"{AGENTS[idx + 1]}_agent",
            "call_count": 900 - idx * 70,
            "avg_latency_ms": 1200 + idx * 330,
            "failed_count": idx * 8,
            "total_tokens": 80000 + idx * 32000,
        }
        for idx in range(len(AGENTS) - 1)
    ]
    return {"metric_date": metric_date.isoformat(), "nodes": nodes, "links": links}


def history_alerts(metric_date: date) -> List[Dict]:
    """
    生成历史批处理告警的测试数据。

    Args:
        metric_date (date): 业务日期

    Returns:
        List[Dict]: 历史告警字典列表
    """
    return [
        {
            "alert_id": f"hist_{metric_date.isoformat()}_{idx}",
            "metric_date": metric_date.isoformat(),
            "alert_type": alert_type,
            "level": level,
            "agent_id": agent,
            "current_value": current,
            "threshold": threshold,
            "source": "batch",
            "status": "open",
        }
        for idx, (alert_type, level, agent, current, threshold) in enumerate(
            [
                ("high_error_rate", "warning", "search_agent", 0.24, 0.20),
                ("high_p95_latency", "warning", "writer_agent", 12400, 10000),
                ("high_cost_task", "info", "analysis_agent", 0.72, 0.50),
            ]
        )
    ]
